[FIX] password_security: Fix password stored and token to reset password invalid

moylop260 opened the pull request:
Steps to reproduce issue:
1. Create a user with a valid password
2. Use the `Forget password` button
3. Use the same previous password

Saw:
Error of history password. Good!
Try to fix it your password and now the token to reset password is invalid. Wrong!
Check the fields user.password and is filled. Wrong!

Expected:
Error of history password and allow fix it and don't store user.password field.

Video:
[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/4PDz2djXuPU/0.jpg)](https://www.youtube.com/watch?v=4PDz2djXuPU)


If you don't want configure a email smtp apply the following patch:
```diff
diff --git a/addons/mail/models/mail_template.py b/addons/mail/models/mail_template.py
index 76d8ae741..316d96a3f 100644
--- a/addons/mail/models/mail_template.py
+++ b/addons/mail/models/mail_template.py
@@ -560,6 +560,7 @@ class MailTemplate(models.Model):
                 updated with given values dict
            :returns: id of the mail.message that was created
         """
+        raise_exception=False
         self.ensure_one()
         Mail = self.env['mail.mail']
         Attachment = self.env['ir.attachment']  # TDE FIXME: should remove dfeault_type from context
```

The query executed from the video is:
`psql openerp_test -c "SELECT login FROM res_users WHERE password != ''"`
`psql openerp_test -c "UPDATE res_users SET password=Null"`

MERGER
=======
Please merge just one commit using merge & squash of github

moylop260 committed:
[FIX] password_security: Fix password stored

lasley reviewed (approved):
Excellent catch @moylop260

yajo commented on password_security/models/res_users.py:
```diff
@@ -81,6 +81,8 @@ def check_password(self, password):
         if not re.search(''.join(password_regex), password):
+            self.env.cr.rollback()
+            self.invalidate_cache()
             raise PassError(_(self.password_match_message()))
```
Whoah, I didn't see this! Remove the call to `_()` here please.

yajo commented on password_security/models/res_users.py:
```diff
@@ -81,6 +81,8 @@ def check_password(self, password):
             password_regex.append(r'(?=.*?\W)')
         password_regex.append('.{%d,}$' % company_id.password_length)
         if not re.search(''.join(password_regex), password):
+            self.env.cr.rollback()
```
I feel all is much more clean with `with self.env.cr.savepoint():`, which rolls back automatically if any exception is found inside the block (the same for all cases below).

moylop260 replied:
Sorry, but I didn't understand.
Where you have planned add a `savepoint`?


- Note: The module `auth_signup` has the `write`:
   
  - [auth_signup/models/res_users.py#L42](https://github.com/odoo/odoo/blob/676ba1bdb/addons/auth_signup/models/res_users.py#L42)

yajo replied:
Oh yes sorry, forget about that.

Anyways, isn't transaction rolled back automatically when an exception is raised?

moylop260 replied:
Yeah, you are right!
I deleted the rollback and it is working just using invalidate_cache

For a strange reason the website frontend is not cleaning the cache.
Because the same exception using backend is working without clean cache.

So, maybe this is a issue related with website but I didn't find a way to reproduce it without a custom module and will be hardly that odoo fixed.

moylop260 replied:
Wait me I didn't test it with the real fail case... sorry.
I will come back with the news

moylop260 replied:
Yeah, I find the real issue.
The method `def _set_password(` don't run a rollback if an exception is raised.
So I just changed the inherited method to write validation
I removed rollback and clear caches

moylop260 committed:
[REF] password_security: use a unified check_password private method to validate rules and history password

moylop260 commented:
@lasley 
I reproduced this issue using frontend `/web/reset_password` and raising the password history exception.

I mean,

| exception raised          | UI                          | Issue detected |
|------------------|-----------------------------|----------------|
| password string rules   | User backend wizard         | No             |
| password string rules   | Frontend web/reset_password | No             |
| password history | User backend wizard         | No             |
| password history | Frontend web/reset_password | Yes            |

So I removed the rollbacks and invalidate_cache and moved the history check to rules string check method.
Thanks @Yajo for comments

Could you review it again, please?

lasley reviewed (approved):
LGTM thanks @moylop260
